### PR TITLE
Opitmization of  layerNorm test

### PR DESCRIPTION
**Pull request by JamesLim-sy**
### PR types

Performance optimization
### PR changes

OPs
### Describe

- Conception 
Theoretically,   the lars op was consist of 2 kernels :
    - L2 norm of grad and param respectively to generate fundamental hyper-param for lars update.
    - Lars update.

- Drawback:
Originally, the L2 norm kernel was implemented with combined eigen kernel below. As shown by codes, a combined L2 norm kernel relied on two eigen kernels. Hence, while training with lars optimizer,  eigen kernel would be called 4 times in total. Take  lars update cuda kernel into account,  5 kernel were needed for only 1 lars op.
```
    (eigen_g.template cast<MPDType>() * rescale_grad).square().sum().sqrt();
```
截屏2021-09-06 下午7 13 34

- Optimization method
 1. It is necessary to combine whole L2 norm calculation into only 1 cuda kernel,  then kernels needed for lars updating would shrink from 5 to 2.
 2. L2 norm is acquired with two steps : 
      -   Partial reduction result of pow(grad, 2) and pow(param, 2) of each thread-block and store into temporary global memory.
      -   In second cuda kernel, read partial reduction results from gloabl memory and reduce to get the final l2 norm result.
 3. More than just split L2 norm calculation, no more than 1024 * sizeof(DataType)  was needed for partial reduction in the first  cuda kernel by design,  in case of too much temporary memory was occupied.
 4. Second cuda kernel mainly works for lars update, and vectorize IO was applied for it to gain the performance.

- Performance 
1. After optimization of lars op, the training performance increase from 2969 ips to 3040 ips. 
2. The total kernel needed shrink from 5 to 2.
image

- Note
Implement with only one kernel to accomplish whole lars op`s function was shown in site below, 
https://github.com/JamesLim-sy/Paddle/commit/352930d949d952886394d68896bcf9d331ceb79a
though uncertain performance change with introduction of compile flag -rdc=ture prevent it from estabulish.

### CODE REVIEW

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 1]
```diff
@@ -14,61 +14,236 @@ limitations under the License. */
...
 
 template <typename T>
 using MultiPrecisionType = typename details::MPTypeTrait<T>::Type;
 
+__device__ __forceinline__ float square_root(float x) { return sqrtf(x); }
+__device__ __forceinline__ double square_root(double x) { return sqrt(x); }
+__device__ __forceinline__ float fma_root(float x, float y, float z) {
+  return fmaf(x, y, z);
+}
+__device__ __forceinline__ double fma_root(double x, double y, double z) {
```
Comment (AnnaTrainingG) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 注意函数命名

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 2]
```diff
@@ -14,61 +14,236 @@ limitations under the License. */
...
     v_out[i] = v_new;
     p_out[i] = static_cast<T>(p_new);
-    if (master_p_out) master_p_out[i] = p_new;
+    master_p_out[i] = p_new;
+  }
+}
+
+template <typename MT, int VesSize>
+__device__ inline void VectorizeLarsUpdate(
+    const MT* __restrict__ g, const MT* __restrict__ v, MT* __restrict__ p_out,
```
Comment (AnnaTrainingG) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这个和上面那个函数只是数据类型不一样？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 3]
```diff
@@ -14,61 +14,236 @@ limitations under the License. */
...
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  int grid_stride = LARS_BLOCK_SIZE * gridDim.x;
+  const MT rescale_grad_pow = rescale_grad * rescale_grad;
+  MT p_val = threadIdx.x < thresh ? p_tmp_buffer[threadIdx.x] : 0;
+  MT g_val = threadIdx.x < thresh ? g_tmp_buffer[threadIdx.x] : 0;
+  __syncthreads();
+  MT p_n = square_root(math::blockReduceSum<MT>(p_val, FINAL_MASK));
+  MT g_n = square_root(rescale_grad_pow *
+                       math::blockReduceSum<MT>(g_val, FINAL_MASK));
+
```
Comment (AnnaTrainingG) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这里不需要加sync 吗？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 4]
```diff
@@ -14,61 +14,236 @@ limitations under the License. */
...
-    } else {
-      InnerCompute<T>(ctx, multi_precision);
-    }
+    InnerCompute(ctx, multi_precision);
   }
 
  private:
-  template <typename MT>
   void InnerCompute(const framework::ExecutionContext& ctx,
                     const bool multi_precision) const {
```
Comment (AnnaTrainingG) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这个函数只调用一次，可以直接放到compute里吧

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 5]
```diff
@@ -14,61 +14,236 @@ limitations under the License. */
...
+  MT g_val = threadIdx.x < thresh ? g_tmp_buffer[threadIdx.x] : 0;
+  __syncthreads();
+  MT p_n = square_root(math::blockReduceSum<MT>(p_val, FINAL_MASK));
+  MT g_n = square_root(rescale_grad_pow *
+                       math::blockReduceSum<MT>(g_val, FINAL_MASK));
+
+  const MT lr = learning_rate[0];
+  MT local_lr = lr;
+  if (lars_weight_decay > static_cast<MT>(0) && p_n > static_cast<MT>(0) &&
+      g_n > static_cast<MT>(0)) {
```
Comment (AnnaTrainingG) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这里的0 每进行一次判断，可能就需要类型转换3次，是否可以合并一下， g_n 和p_n都是非负吧，q_n * p_n* lars_weight_decay > 0 ?

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 2 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: `VectorizeLarsUpdate`和`VectorizeLarsUpdateMP`拆分开的主要原因是,
`VectorizeLarsUpdateMP`服务于fp16 和 fp32的混合精度计算，所以可以确定每次读写4个数据，另外结果要同时写出fp6 output和 fp32 master_output。而`VectorizeLarsUpdateMP`服务于fp32或者fp64的计算，且结果不需要另外写出master_output。可以考虑用仿函数等方法合并，下个commit 比较合并前后的性能数据。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 5 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这块的判断确实可以压缩掉，忘记这一茬了，下个commit修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 4 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 好的，下个commit修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 1 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 下个commit修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 3 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: math::blockReduceSum<MT>(g_val, FINAL_MASK) 执行完毕之后自带一次sync__ ，且返回的结果应该在寄存器侧，直接求计算不需要带sync

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 6]
```diff
@@ -14,72 +14,256 @@ limitations under the License. */
...
+#define LARS_BLOCK_SIZE 512
+#endif
 
 namespace paddle {
 namespace operators {
 
 template <typename T>
 using MultiPrecisionType = typename details::MPTypeTrait<T>::Type;
 
+__device__ __forceinline__ float SquareRoot(float x) { return sqrtf(x); }
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 函数里面为啥要加个`Root`呢？直接叫`Square`、`Fma`不也OK？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 7]
```diff
@@ -14,72 +14,256 @@ limitations under the License. */
...
+    VecMType p_data = master_p_vec[i];
 
-  if (lars_weight_decay > static_cast<MT>(0) && p_n > static_cast<MT>(0) &&
-      g_n > static_cast<MT>(0)) {
-    local_lr =
-        lr * lars_coeff * p_n / (g_n + lars_weight_decay * p_n + epsilon);
+#pragma unroll
+    for (int j = 0; j < 4; ++j) {
+      MT grad = static_cast<MT>(g_data.val[j]) * rescale_grad;
+      v_new.val[j] =
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 可以直接用`v_new[i]`，`[]`运算符有重载了。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 8]
```diff
@@ -14,72 +14,256 @@ limitations under the License. */
...
-    T* p_out, MT* v_out, const MT epsilon, const MT* master_p, MT* master_p_out,
-    const MultiPrecisionType<T> rescale_grad) {
-  const MT lr = static_cast<MT>(learning_rate[0]);
-  MT local_lr = lr;
-  const MT p_n = static_cast<MT>(p_norm[0]);
-  const MT g_n = static_cast<MT>(g_norm[0]);
+__device__ inline void VectorizeLarsUpdateMP(
+    const T* __restrict__ g, const MT* __restrict__ v, T* __restrict__ p_out,
+    MT* __restrict__ v_out, const MT* __restrict__ master_p,
+    MT* __restrict__ master_p_out, const MT mu, MT local_lr,
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: `master_p`和`master_p_out`是同一个指针吧？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 2 above)
Comment (Xreki) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 两个函数的主要区别应该是有没有`master_p`，可以在Kernel中通过`master_p`是否为`nullptr`区分。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 9]
```diff
@@ -14,72 +14,256 @@ limitations under the License. */
...
+    local_lr = lr * lars_coeff * p_n /
+               (FmaRoot(lars_weight_decay, p_n, g_n) + epsilon);
+  }
+
+  if (master_p) {
+    VectorizeLarsUpdateMP<T, MT>(g, v, p_out, v_out, master_p, master_p_out, mu,
+                                 local_lr, lars_weight_decay, rescale_grad, tid,
+                                 grid_stride, numel);
+  } else {
+    if (sizeof(T) < 64) {  // 64 stands for sizeof(double)
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 不用考虑首地址是否满足对齐要求了吗？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 10]
```diff
@@ -14,72 +14,256 @@ limitations under the License. */
...
+    const MT* __restrict__ learning_rate, MT* __restrict__ p_tmp_buffer,
+    MT* __restrict__ g_tmp_buffer, const MT mu, const MT lars_coeff,
+    const MT lars_weight_decay, const MT epsilon, const MT rescale_grad,
+    const int thresh, const int64_t numel) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  int grid_stride = LARS_BLOCK_SIZE * gridDim.x;
+  const MT rescale_grad_pow = rescale_grad * rescale_grad;
+  MT p_val = threadIdx.x < thresh ? p_tmp_buffer[threadIdx.x] : 0;
+  MT g_val = threadIdx.x < thresh ? g_tmp_buffer[threadIdx.x] : 0;
+  __syncthreads();
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 前面这一段代码的功能，加些注释说明下。

**Pull request closed without merging**